cachi2/core/package_managers/pip.py

```diff
@@ -1411,12 +1423,29 @@ def _download_dependencies(
 
         if require_hashes or req.kind == "url":
             hashes = req.hashes or [req.qualifiers["cachito_hash"]]
-            _verify_hash(download_info["path"], hashes)
+            _verify_hash(download_info["path"], list(map(_to_checksum_info, hashes)))
 
         download_info["kind"] = req.kind
-        downloads.append(download_info)
+        downloaded.append(download_info)
+
+    if allow_binary:
+        log.info("Downloading %d wheel(s) ...", len(to_download))
+        files: dict[str, Union[str, PathLike[str]]] = {pkg.url: pkg.path for pkg in to_download}
+        asyncio.run(async_download_files(files, get_config().concurrency_limit))
+
+        for pkg in to_download:
+            try:
+                key = (canonicalize_name(pkg.name), canonicalize_version(pkg.version))
+                user_req_checksums = user_checksums.get(key, set())
+                checksums = pkg.checksums if pkg.checksums else user_req_checksums
+                _verify_hash(pkg.path, checksums)
+            except PackageRejected:
+                pkg.path.unlink()
+                log.warning("The file is removed from the output directory")
 
-    return downloads
+        log.info("Any existing wheels were not re-downloaded")
```

Comment:
I don't think this log message is important (and users probably won't really know what it means)